db/db_logical_block_size_cache_test.cc

```diff
@@ -0,0 +1,366 @@
+
+    // Delete one handle won't drop the cache.
+    ASSERT_OK(db->DestroyColumnFamilyHandle(cfs[0]));
+    ASSERT_EQ(20, cache_->GetLogicalBlockSize(cf_path_0_ + "/sst2", 400));
+    ASSERT_EQ(2, ncall_);
+
+    // Delete the last handle will drop the cache.
+    ASSERT_OK(db->DestroyColumnFamilyHandle(cfs[1]));
+    ASSERT_EQ(100, cache_->GetLogicalBlockSize(cf_path_0_ + "/sst", 100));
+    ASSERT_EQ(3, ncall_);
+
+    ASSERT_OK(db->DestroyColumnFamilyHandle(cfs[2]));
+    // db is not deleted yet, so dbname_ is still cached.
+    ASSERT_EQ(0, cache_->GetLogicalBlockSize(dbname_ + "/sst1", 400));
+    ASSERT_EQ(3, ncall_);
+
+    // db is deleted, cache for dbname_ is dropped.
+    delete db;
+    ASSERT_EQ(100, cache_->GetLogicalBlockSize(dbname_ + "/sst", 100));
+    ASSERT_EQ(4, ncall_);
+  }
+}
+
+TEST_F(DBLogicalBlockSizeCacheTest, DestroyColumnFamilyHandle) {
+  // Tests that destroy column family without dropping won't drop the cache,
+  // because compaction and flush might still need to get logical block size
+  // when opening new files.
+  Options options;
+  options.create_if_missing = true;
+  options.env = env_.get();
+
+  DB* db;
+  ASSERT_OK(DB::Open(options, dbname_, &db));
+  ColumnFamilyOptions cf_options;
+  cf_options.cf_paths = {{cf_path_0_, 1024}};
+  ColumnFamilyHandle* cf = nullptr;
+  ASSERT_OK(db->CreateColumnFamily(cf_options, "cf", &cf));
+  // cf_path_0_ is cached.
+  ASSERT_EQ(2, ncall_);
+  ASSERT_EQ(20, cache_->GetLogicalBlockSize(cf_path_0_ + "/sst", 100));
+  ASSERT_EQ(2, ncall_);
+
+  // Delete handle won't drop cache.
+  ASSERT_OK(db->DestroyColumnFamilyHandle(cf));
+  ASSERT_EQ(20, cache_->GetLogicalBlockSize(cf_path_0_ + "/sst1", 200));
+  ASSERT_EQ(2, ncall_);
+
+  delete db;
+
+  ColumnFamilyOptions default_cf_options;
+  default_cf_options.cf_paths = {{dbname_, 2048}};
+
+  std::vector<ColumnFamilyHandle*> cfs;
+  for (int i = 0; i < 2; i++) {
+    ncall_ = 0;
+    if (!i) {
+      printf("Open\n");
+      ASSERT_OK(DB::Open(options, dbname_,
+          {{"cf", cf_options},
+          {"default", default_cf_options}},
+          &cfs, &db));
+    } else {
+      printf("OpenForReadOnly\n");
+      ASSERT_OK(DB::OpenForReadOnly(options, dbname_,
+          {{"cf", cf_options},
+          {"default", default_cf_options}},
+          &cfs, &db));
+    }
+    // cf_path_0_ and dbname_ are cached.
+    ASSERT_EQ(2, ncall_);
+    ASSERT_EQ(20, cache_->GetLogicalBlockSize(cf_path_0_ + "/sst", 100));
+    ASSERT_EQ(2, ncall_);
+    ASSERT_EQ(0, cache_->GetLogicalBlockSize(dbname_ + "/sst", 101));
+    ASSERT_EQ(2, ncall_);
+
+    // Deleting handle won't drop cache.
+    ASSERT_OK(db->DestroyColumnFamilyHandle(cfs[0]));
+    ASSERT_EQ(20, cache_->GetLogicalBlockSize(cf_path_0_ + "/sst1", 200));
+    ASSERT_EQ(2, ncall_);
+    ASSERT_OK(db->DestroyColumnFamilyHandle(cfs[1]));
+    ASSERT_EQ(0, cache_->GetLogicalBlockSize(dbname_ + "/sst1", 201));
+    ASSERT_EQ(2, ncall_);
+
+    delete db;
+  }
+}
```

Comment:
I know multiple DB test is harder to write, but if you are not 100% sure about the cache when dealing with multiple DBs, I encourage you to write one test case for multiple DBs.

Reply:
added tests for situations for multiple DBs.

Reply:
added two more tests regarding multiple DBs.